### ISSUE TITLE
eibwrite-cgi is using different PAs

### ISSUE THREAD

**Issue by ChristianMayer** [knxd/knxd#495]
Running repeatedly `http://webserver.local:38080/cgi-bin/w?s=SESSION&a=12/7/2&v=80` (or `export REQUEST_METHOD=GET QUERY_STRING='s=SESSION&a=12/7/90&v=80152e' && /usr/lib/cgi-bin/w` from the command line where `/usr/lib/cgi-bin/w` is `eibread-cgi`) results in using a different physical address for each request.

This might also be the reason why these calls were failing initially with a `Connect failed` error (or a `Allocate: no more free addresses!` in the knxd log) when I gave it only very few client-addrs (it was only one or two)

eibread-cgi and eibwrite-cgi should always use the same PA as they are one (virtual) KNX device

**Comment by smurfix** [comment]
Client addresses should be freed as soon as the CGI process ends. "Connect failed" looks like knxd wasn't even running?

These CGI programs use the standard knxd interface. knxd has no way to know that the connection is from them. Using the addresses in a round-robin fashion is a feature because that prevents program B from getting confused when a late reply to program A arrives after A has ended.

I don't consider them one virtual KNX device. You don't know how many programs use them to talk to knxd. In any case every time they get started it's a new connection as far as knxd is concerned.

Assuming you allocate more than two client addresses, does that cause an actual problem, or is this just your sense of aesthetics? ;-)

**Comment by ChristianMayer** [comment]
Yes, knxd was running, otherwise I would get an `Open failed` error, I guess.

The `eibwrite-cgi` works in a fire and forget fashion. There is no reply from the bus, so nothing can be confused.
Each time the visu wants to write something to the knx it starts a new `eibwrite-cgi` (we are talking CGI here and not FastCGI, so the processes are started and ended all the time)

The visu is a virtual knx device for me. Giving it multiple PAs would be very confusing when looking at a bus monitor.
Using a bus monitor usually is when you are doing real work and don't want to be distracted by confusing output. So it's not just aesthetics.
For bigger installations with multiple visus open at the same time (e.g. office building) it might even become an issue where you will be running out of addresses when you need different ones for each call.

**Comment by smurfix** [comment]
Well … if you want to run more than one visu you need to associate replies to the correct client, so each CGI process still needs a different address. If you want to change that and have a permanent association you'll need to write a long-running CGI handler, and you need to teach your clients to keep the HTTP connection open.

The latter is not knxd's problem. The former is certainly possible, but IMHO not a particularly good idea given the current abysmally-undocumented state of the internal knxd "protocol".

My current plan is to keep the protocol out of knxd entirely; the idea is to rewrite the HTTP frontend in Python, along with the group message cache and a reasonable subset of the internal protocol. I hope to have some code for that sometime early next year.

**Comment by ChristianMayer** [comment]
The web interface with eibread-cgi and eibwrite-cgi is (on purpose!) completely stateless.
When the Visu does a write it doesn't need a reply, so the only reason it needs an address is that the KNX bus protocol is designed in such a way that the sender PA is known on the bus.
The only time a reply is needed is when a read to a GA unknown to the knxd cache is happening. But even then there's nothing that is remembered on the visu, the eibread-cgi/eibwrite-cgi or the knxd side. Just sending the read request (fire and forget) is enough as the answer will come just like an ordinary write from the requested knx device.

All of this allows the full server installation of the knxd and web server for the visu to run on very small hardware (like a 15 years old router) - which is for the required performance absolutely sufficient and still gives the user a real time performance. Adding python as a dependency would be completely contradictory to such a light weight system. 

Of course this could be "enhanced" to a statefull solution that is remembering the clients and using long running processes (and thus most likely FastCGI). The CometVisu protocol is actually prepared for that by using a login call as well as a session ID. It could also help to reduce the traffic by not always repeating the GAs of interest.
As this doesn't bring any relevant benefit but has the chance of introducing many bugs and even breaks the advantage of being light weight it didn't seem attractive to follow that route so far.

Trying to use the latest knxd instead of the rock solid 0.0.5.1 has already cost me 2-3 days that I could have also invested to finally implement the MQTT backend. And it's still not finished as e.g. the new knxd needs so long to compile that the Travis CI cross compile job for the ARM systems is timing out and I still have no solution for that.
Currently I'm wondering whether I should still invest time into this version update when it seems that the development direction of the knxd is heading the wrong way.

**Comment by dallmair** [comment]
knxd v0.12 saw the introduction of the `single` filter. I tested that in the current version and can confirm that this filter applied to the unix socket configuration of knxd causes it to send out all packets with knxd's PA instead of using any of the additionally configured client PAs, like eibd did. So if you think of your knxd instance as being "one (virtual) KNX device", you can configure it to behave as such. Therefore, I'm closing this issue now.
